runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -9065,6 +9065,145 @@ inlineNanoTime(
 #endif
 #endif // LINUX
 
+TR::Register* J9::X86::TreeEvaluator::inlineMathFma(TR::Node* node, TR::CodeGenerator* cg)
+   {
+   TR::Node *firstChild = node->getFirstChild();
+   TR::Node *secondChild = node->getSecondChild();
+   TR::Node *thirdChild = node->getThirdChild();
+
+   TR::Register *lhsReg = NULL;
+   TR::Register *midReg = NULL;
+   TR::Register *rhsReg = NULL;
+   TR::Register *result = cg->allocateRegister(TR_FPR);
+
+   bool memLoadLhs = !(firstChild->getRegister() || firstChild->getReferenceCount() != 1 ||
+                       !firstChild->getOpCode().isLoadVar());
+
+   bool memLoadMiddle = !(secondChild->getRegister() || secondChild->getReferenceCount() != 1 ||
+                          !secondChild->getOpCode().isLoadVar());
+
+   bool memLoadRhs = !(thirdChild->getRegister() || thirdChild->getReferenceCount() != 1 ||
+                       !thirdChild->getOpCode().isLoadVar());
+
+   bool is64Bit = getNodeIs64Bit(node, cg);
+
+   TR::InstOpCode::Mnemonic fpMovRegRegOpcode = is64Bit ? TR::InstOpCode::MOVSDRegReg : TR::InstOpCode::MOVSSRegReg;
+   result->setIsSinglePrecision(!is64Bit);
+
+   TR_ASSERT_FATAL(cg->comp()->target().cpu.supportsFeature(OMR_FEATURE_X86_FMA), "Cannot generate inline fma implementation without FMA extensions");
+
+   // Choose fma instruction carefully, based on operand form, to reduce number of copies
+   if (memLoadLhs)
+      {
+      TR::InstOpCode::Mnemonic opcode = is64Bit ? TR::InstOpCode::VFMADD231SDRegRegMem : TR::InstOpCode::VFMADD231SSRegRegMem;
+      // swap a and c; fma = b * c + a
+
+      TR::MemoryReference *lhsMR = generateX86MemoryReference(firstChild, cg);
+      midReg = cg->evaluate(secondChild);
+      memLoadMiddle = false; // No choice but to evaluate;
```

Comment:
Suppose `memLoadLhs` and `memLoadMiddle` are both true, but `memLoadRhs` is false.  Would it be better to use `VFMADD132S{S|D}`, loading the value of the second child into the `result` register?  I think that would avoid an extra copy of `rhsReg` into the `result` register.